File: back-end/src/main/java/pl/straburzynski/campfiresongs/service/SongConverter.java
package pl.straburzynski.campfiresongs.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import pl.straburzynski.campfiresongs.exception.JsonParseException;
import pl.straburzynski.campfiresongs.model.Song;
import pl.straburzynski.campfiresongs.model.SongDto;

@Service
@RequiredArgsConstructor
public class SongConverter {

    private final ObjectMapper objectMapper;

    public String stringifyFromSongDto(SongDto songDto) {
        try {
            return objectMapper.writeValueAsString(songDto);
        } catch (JsonProcessingException e) {
            throw new JsonParseException("stringifyFromSongDto");
        }
    }

    public Song convertFromSongDto(SongDto songDto) {
        return objectMapper.convertValue(songDto, Song.class);
    }

    public SongDto parseFromSongData(String songData) {
        try {
            return objectMapper.readValue(songData, SongDto.class);
        } catch (JsonProcessingException e) {
            throw new JsonParseException("parseFromSongData");
        }
    }

    public SongDto convertFromSong(Song song) {
        return objectMapper.convertValue(song, SongDto.class);
    }
}

File: back-end-kotlin/src/main/kotlin/pl/straburzynski/campfiresongs/song/service/SongConverter.kt
package pl.straburzynski.campfiresongs.song.service

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.stereotype.Service
import pl.straburzynski.campfiresongs.song.exception.JsonParseException
import pl.straburzynski.campfiresongs.song.model.Song
import pl.straburzynski.campfiresongs.song.model.SongDto


@Service
class SongConverter(private val objectMapper: ObjectMapper) {

    fun stringifyFromSongDto(songDto: SongDto?): String = try {
        objectMapper.writeValueAsString(songDto)
    } catch (e: JsonProcessingException) {
        throw JsonParseException("stringifyFromSongDto")
    }

    fun convertFromSongDto(songDto: SongDto?): Song =
        objectMapper.convertValue(songDto, Song::class.java)

    fun parseFromSongData(songData: String?): SongDto = try {
        objectMapper.readValue(songData, SongDto::class.java)
    } catch (e: JsonProcessingException) {
        throw JsonParseException("parseFromSongData")
    }

    fun convertFromSong(song: Song?): SongDto =
        objectMapper.convertValue(song, SongDto::class.java)
}